File: redis_chat_app.py
import redis
import threading

client = redis.StrictRedis(host='localhost', port=6379, decode_responses=True)


def listen_for_messages(room, username):
    pubsub = client.pubsub()
    pubsub.subscribe(room)

    print(f"Joined chat room '{room}'. Waiting for messages...\n")

    for message in pubsub.listen():
        if message['type'] == 'message':
            if not message['data'].startswith(f"{username}:"):
                print(f"{message['data']}")


def send_message(room, username):
    while True:
        message = input(f"[{username}@{room}]> ")
        if message.lower() == 'exit':
            print(f"Exiting chat room '{room}'...")
            break
        client.publish(room, f"{username}: {message}")


def menu():
    username = input("Enter your username: ")
    room = input("Enter chat room name: ")

    listener_thread = threading.Thread(target=listen_for_messages, args=(room, username))
    listener_thread.daemon = True
    listener_thread.start()

    send_message(room, username)


if __name__ == '__main__':
    menu()

File: chat/RedisChatApp.java
import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisPubSub;
import java.util.Scanner;

public class RedisChatApp {

    public static void main(String[] args) {
        Jedis publisherJedis = new Jedis("localhost", 6379);
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter your username: ");
        String username = scanner.nextLine();
        System.out.print("Enter chat room name: ");
        String room = scanner.nextLine();

        Thread subscriberThread = new Thread(() -> {
            Jedis subscriberJedis = new Jedis("localhost", 6379);

            JedisPubSub jedisPubSub = new JedisPubSub() {
                @Override
                public void onMessage(String channel, String message) {
                    if (!message.startsWith(username + ":")) {
                        System.out.println(message);
                    }
                }
            };

            System.out.println("Joined chat room '" + room + "'. Waiting for messages...\n");
            subscriberJedis.subscribe(jedisPubSub, room);
        });

        subscriberThread.start();

        while (true) {
            System.out.print("[" + username + "@" + room + "]> ");
            String message = scanner.nextLine();
            if (message.equalsIgnoreCase("exit")) {
                System.out.println("Exiting chat room '" + room + "'...");
                break;
            }

            publisherJedis.publish(room, username + ": " + message);
        }

        publisherJedis.close();
        scanner.close();
    }
}